### PR TITLE
Lego_teamBot_jungleLib_Submission

### PR DESCRIPTION
**Pull request by Kairus101**
This was to make lego submittable, along with teamBotBrain.
Lego: Integrated teamBotBrain, added dynamic item builds, added laning capability.
TeamBotBrain: Chose a guessing method, added a check for valid lanes, optimized algorithm, added  jungle lane, removed tie-break, added player jungle check, omitted lanes to increase performance.
Core: changed GetLaneBreakdown to show when a player isn't really in a lane at all.

### CODE REVIEW

**Comment by eeyrjmr** [comment]
you could have left pull request #9 and edited your fork. It was still reviewable
ONLY upon a merge does the present content of your fork get pulled in

having another pull request for the teambot changes would have been cleaner. #9 could then have been dependant on #12 being competed.

File: bots/core.lua [hunk 1]
```diff
@@ -1548,24 +1548,29 @@ function core.GetLaneBreakdown(unit)
...
-		botDist = Vector3.Distance2D(position, vBotPoint)
+	local vecBotPoint = core.GetFurthestPointOnPath(position, metadata.GetBottomLane(), core.bTraverseForward)
+	if vecBotPoint then
+		botDist = Vector3.Distance2D(position, vecBotPoint)
 	end
 	
 	--pick two lowest ones
 	local nBiggestDist = max(topDist, max(midDist, botDist))
 	
+	local nLowestDist = max(topDist, max(midDist, botDist))
```
Comment (Kairus101) [inline bots/core.lua]: Should be:
local nLowestDist = min(topDist, midDist, botDist)

File: bots/teambot/teambotbrain.lua [hunk 2]
```diff
@@ -1044,273 +1244,91 @@ function object:BuildLanes()
...
-			if core.NumberElements(tExposedLane) < 2 then
-				tLaneToFill = tExposedLane
-				if bDebugEchos then print(" in the Exposed lane\n") end
-			elseif core.NumberElements(tSafeLane) < 2 then
-				tLaneToFill = tSafeLane
-				if bDebugEchos then print(" in the Safe lane\n") end
+	-- /Tutorial
+
+	object.lanePreferences={}
+	nHighestCombo=0
```
Comment (Kairus101) [inline bots/teambot/teambotbrain.lua]: Should be:
nHighestCombo = -500

**Comment by eeyrjmr** [comment]
The teambot has fixed the initial concern so thats good. Need to guage the specifics of the teambot to try to get that in ASAP for general usage.

ok a couple of things

1) with the teambot the bots have a better idea about who is going where - good. For the legio's that goto lane, starting with blight+2xbuckler still valid? is it worth concidering multiple item build routes?

2) the bot's will stand by the weak camp in the jungle until just before 1:00 in an attempt to stack the camp. This might just be downto them being antler this time, but it failed to stack - might be downto antlers, I know there is one creep type that is a pig to stack against

3) seems a bug somewhere in the code triggering a core.lua:1568  attempt to call global 'min' (a nil value)

**Comment by Kairus101** [comment]
1. Multiple item builds are supported and programmed in - simply commented out. It requires advanced shopping lib, then it can be un-commented and boom, multiple item builds depending on lane.
2. The stack normally works, but healers are always a pain to stack for. I would have to add lots of code in order to take that one camp into account.
3. Good point, for some reason it didn't show that it needed to be changed, it's just the the min hasn't been declared at the top, I'll fix this asap (most likely by the time you see this comment)

**Comment by eeyrjmr** [comment]
4) due to the fact the bots kind of sit there and just accrue gold, the inevitable "the bot has 4k of gold" 
Now due to the nature of how the bots just iterate over the item list, the bot will go to complete helm --> ghostmarchers... 
In practice if someone had 4k gold and their core item was a pk, they would probably pick that up before completing ghost marchers simply because this is the main item for legio

**Comment by Kairus101** [comment]
4) Not too sure what you mean. You think he should save gold until he has 2150 then buy pkey? If he dies beforehand or something then get ghost marchers instead? The other thing is that ghost marchers enhance farming too, and he does get the pkey quite quickly after the ghost marchers

**Comment by eeyrjmr** [comment]
not just outright save for pkey, but with the present bot behaviour where they don't return to base UNLESS really hurt or dead... the bot could be sitting on alot of gold. human players, when faced with such a large sum would buy a pk.
Saying that human players wouldn't build up 4k of gold :)

**Comment by eeyrjmr** [comment]
Well core.lua changes seem valid. The bots I have tried do not seem to be adversely effected by it.
It would have been better (as I mentioned) if the core.lua changes were themselves their own pull request (equally teambot... leaving jungle+legio as its own...) 

I will fork the tree, merge in this core.lua and initiate a pull request on that to get the core in.

teambot seems to have some arbitary designation as to what is a carry is

if nAttackRange > 500 then --long range, most likely a carry, ex: Flint, Emerald Warden, Valkyrie
if nAttackRange >= 450 then --long range int, most likely a semi-carry, pyro, aluna, rhaps, myrm,

Would be better if the S2 designated rolltypes was exposed to the bots. I have requested this from Malloc.

So far though so good with Legio.

**Comment by Kairus101** [comment]
Ok, core.lua changes nothing other than something that is needed for teamBotBrain, so, I wouldn't expect to see any effect on other bots.

w.r.t saving for a portalkey, lego WILL return to base when he has lots of gold/low hp/low mana, this is added in the healAtWell behavior. Also, he is rigged to use advanced shopping when it gets approved, and with his earlier items, his farming gets much better.

w.r.t the designation as to what is a carry is, they come in all shapes and sizes. But I looked through the entire hero pool and these were the most accurate numbers I could get, it has a few odd ones, but, in theory once everyone makes their bot have lane preferences, this won't be a problem.

**Comment by eeyrjmr** [comment]
This is why I asked malloc to expose what S2 classes a hero as since this something that is alot more quantifiable as to what is a carry, what is a support. 

I can see this, but for some reason I was seeing instances where the bot wasn't returning, ie this utility is still quite low. 4000gold only adds +16 to a utility which could be almost zero if the health is high and the mana is high.

--Return to well, based on more factors than just health.
function HealAtWellUtilityOverride(botBrain)
    return object.HealAtWellUtilityOld(botBrain)_1.75+(botBrain:GetGold()_8/2000)+ 8-(core.unitSelf:GetManaPercent()*8)
end
object.HealAtWellUtilityOld = behaviorLib.HealAtWellBehavior["Utility"]
behaviorLib.HealAtWellBehavior["Utility"] = HealAtWellUtilityOverride

This is more of a botbrain feature since bots REALLY need to use stash for some big items. Think what a human would do "I am legio, wow I have saved 2000g, 150g more and I have pk, sod the other stuff"

**Comment by Kairus101** [comment]
If the bot is full hp and full mana, he shouldn't really be returning just yet, instead, when advanced shopping is in the game, a courier will bring his items to him. This isn't a "We have lots of gold, lets go back" thing, it's a "We have no mana, lots of gold, why push or jungle, I won't be very effective" thing. Still unsure on the rushing pkey, whether lego would be able to farm it quickly enough to have an impact is questionable, but I believe he needs the hotbl in order to farm, as players, we are better at farming and kiting than bots, but legoBot wouldn't survive that long in jungle without it.

**Comment by eeyrjmr** [comment]
nono, you misunderstood. I am not implying that the bot should rush a pk (at the very least the skill and item's a bot buys is upto your). My point was the bots were sitting on a relative goldmine with not that much incentive to spend the gold (my pyrobot which I use to test stuff would do the same thing, sitting on 6k...)

As mentioned this is an inherent flaw in all the bots and partly why I wrote that spending function to force the bot back to the fountain based upon the amount of gold he has AND the total value of the items in the present table (basically a slightly more adv version of what you have) 

Now if the idea is to have it solved by advShopping then so be it. is that a blue sky's submission or an in progress?

**Comment by Kairus101** [comment]
I'm reviewing it now, It's working well and I'm about 1/2 done, and lego works much better with it.

**Comment by eeyrjmr** [comment]
I still don't like this method of deciding what a carry is

Arti has a range of 450, mq even lower 

I know you say bots can hardcode the lane it wants, but what if 5 bots are chosen and they all want ... bottom, it then defer to this logic to decide. then there are the bots that don't have this added OR is the idea to bump or present bots to provide it with a prefered lane?

**Comment by Kairus101** [comment]
I will add laning code to each bot that is currently implemented. It is impossible to classify what role a hero should be from the small amount of information given, so, we generalise. Yes, of course it is wrong for some heroes, but it is a failsafe.

**Comment by eeyrjmr** [comment]
Can you merge against COMMUNITY for this pull request.
I have locally merged w.r.t. community to test this bot against present 

--edit-- seems it is working now, must have been something up with my git pull and the initial patch or something

The junglelib might need a bit more work. the oposition just had glaciusBot go suicide long so legio could jungle and then a dual short.

Equally legio when he is jungling seems to ignore any other threats... I followed him around for a good min and he responded like I wasn't there and even when I started attacking him he priorities attacking the creeps before fleeing home when his hp went too low.

The console was getting spammed with alot of warnings but unsure what bot that was todo with

**Comment by Kairus101** [comment]
The junglelib might need a bit more work. the oposition just had glaciusBot go suicide long so legio could jungle and then a dual short.
Fixed by adding laning prefs to Glacius.

Equally legio when he is jungling seems to ignore any other threats... I followed him around for a good min and he responded like I wasn't there and even when I started attacking him he priorities attacking the creeps before fleeing home when his hp went too low.
This is a problem with lego it seems - I will fix this one soon, but I think we should work on teamBot, as it doesn't have any dependencies like Lego does.

**Pull request closed without merging**